crates/witness-generator/src/main.rs

```diff
@@ -1,4 +1,168 @@
-fn main() {
-    // TODO: generate execution witness and save to file at workspace level.
-    // Call the directory zkevm-fixtures-witnesses
+use anyhow::{Context, Result, anyhow};
+use clap::{Parser, Subcommand};
+use std::path::{Path, PathBuf};
+use witness_generator::{
+    WitnessGenerator,
+    eest_generator::ExecSpecTestBlocksAndWitnessBuilder,
+    rpc_generator::{RpcBlocksAndWitnessesBuilder, RpcFlatHeaderKeyValues},
+};
+
+#[derive(Parser)]
+#[command(name = "zkvm-fixture-generator")]
+#[command(about = "Generate fixtures for zkEVM benchmarking tool")]
+#[command(version)]
+struct Cli {
+    /// Output folder for generated fixtures
+    #[arg(short, long, default_value = "zkevm-fixtures-input")]
+    output_folder: PathBuf,
+
+    /// Source of blocks and witnesses
+    #[command(subcommand)]
+    source: SourceCommand,
+}
+
+#[derive(Subcommand, Clone, Debug)]
+enum SourceCommand {
+    /// Generate fixtures from execution specification tests
+    Tests {
+        /// EEST release tag to use (e.g., "v0.1.0"). If empty, the latest release will be used.
+        #[arg(short, long)]
+        tag: Option<String>,
+
+        /// Include only test names containing the provided strings.   
+        #[arg(short, long)]
+        include: Option<Vec<String>>,
+
+        /// Exclude all test names containing the provided strings.
+        #[arg(short, long)]
+        exclude: Option<Vec<String>>,
+    },
+    /// Generate fixtures from an RPC endpoint
+    Rpc {
+        /// Number of last blocks to pull
+        #[arg(long, conflicts_with = "block")]
+        last_n_blocks: Option<usize>,
+
+        /// Specific block number to pull
+        #[arg(long, conflicts_with = "last_n_blocks")]
+        block: Option<u64>,
+
+        /// RPC URL to use (mandatory)
+        #[arg(long)]
+        rpc_url: String,
+
+        /// Optional RPC headers to use (format: "Key:Value")
+        #[arg(long)]
+        rpc_header: Option<Vec<String>>,
+    },
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let cli = Cli::parse();
+
+    println!("Generating fixtures in folder: {:?}", cli.output_folder);
+    if !cli.output_folder.exists() {
+        std::fs::create_dir_all(&cli.output_folder)
+            .with_context(|| format!("Failed to create output folder: {:?}", cli.output_folder))?;
```

Comment:
Going a bit more formal in proper err wrapping in general since got in some rabbit holes by not knowing causes of error messages